sdk/core/src/http_client.rs

```diff
@@ -0,0 +1,126 @@
+use crate::errors::{AzureError, UnexpectedHTTPResult};
+use async_trait::async_trait;
+use http::{Request, Response, StatusCode};
+use hyper::{self, body, Body};
+use hyper_rustls::HttpsConnector;
+
+#[async_trait]
+pub trait HttpClient: Send + Sync + std::fmt::Debug {
+    async fn execute_request(
+        &self,
+        request: Request<&[u8]>,
+    ) -> Result<Response<Vec<u8>>, Box<dyn std::error::Error + Sync + Send>>;
+
+    async fn execute_request_check_status(
+        &self,
+        request: Request<&[u8]>,
+        expected_status: StatusCode,
+    ) -> Result<Response<Vec<u8>>, Box<dyn std::error::Error + Sync + Send>> {
+        let response = self.execute_request(request).await?;
+        if expected_status != response.status() {
+            Err(Box::new(AzureError::from(UnexpectedHTTPResult::new(
+                expected_status,
+                response.status(),
+                std::str::from_utf8(response.body())?,
+            ))))
+        } else {
+            Ok(response)
+        }
+    }
+
+    async fn execute_request_check_statuses(
+        &self,
+        request: Request<&[u8]>,
+        expected_statuses: &[StatusCode],
+    ) -> Result<Response<Vec<u8>>, Box<dyn std::error::Error + Sync + Send>> {
+        let response = self.execute_request(request).await?;
+        if !expected_statuses
+            .iter()
+            .any(|expected_status| *expected_status == response.status())
+        {
+            if expected_statuses.len() == 1 {
+                Err(Box::new(AzureError::from(UnexpectedHTTPResult::new(
+                    expected_statuses[0],
+                    response.status(),
+                    std::str::from_utf8(response.body())?,
+                ))))
+            } else {
+                Err(Box::new(AzureError::from(
+                    UnexpectedHTTPResult::new_multiple(
+                        expected_statuses.to_vec(),
+                        response.status(),
+                        std::str::from_utf8(response.body())?,
+                    ),
+                )))
+            }
+        } else {
+            Ok(response)
+        }
+    }
+}
+
+pub static EMPTY_BODY: [u8; 0] = [];
+
+#[async_trait]
+impl HttpClient for hyper::Client<HttpsConnector<hyper::client::HttpConnector>> {
+    async fn execute_request(
+        &self,
+        request: Request<&[u8]>,
+    ) -> Result<Response<Vec<u8>>, Box<dyn std::error::Error + Sync + Send>> {
+        let mut hyper_request = hyper::Request::builder()
+            .uri(request.uri())
+            .method(request.method());
+
+        for header in request.headers() {
+            hyper_request = hyper_request.header(header.0, header.1);
+        }
+
+        let body = String::from_utf8(request.body().to_vec())?;
+        let hyper_request = hyper_request.body(Body::from(body))?;
+
+        let hyper_response = self.request(hyper_request).await?;
```

Comment:
I've tried but it seems to return an `http::Response<hyper::Body>` . The trait expect an `http::Response<Vec<u8>>`  instead. In my understanding are basically the same thing (ie `Body` implements `From<Vec<u8>>`) so the move should be cheap:

```
error[E0308]: mismatched types
  --> sdk/core/src/http_client.rs:83:12
   |
83 |         Ok(hyper_response)
   |            ^^^^^^^^^^^^^^ expected struct `Vec`, found struct `hyper::Body`
   |
   = note: expected struct `http::Response<Vec<u8>>`
              found struct `http::Response<hyper::Body>`
```

Reply:
I think in this case you can just use `.map` to change the body type instead of constructing a whole new response object https://docs.rs/hyper/0.13.9/hyper/struct.Response.html#method.map

Reply:
That's a great idea! 
I've tried it but unfortunately `Hyper`'s `Body` is a stream so in order to obtain the Vec we must await on `body::to_bytes()`. 
Something like (this is as far as I went, it's not working): 

```rust
Ok(hyper_response.map(async move |b| body::to_bytes(b).await?.to_vec()))
```

But: 
1. And async closures are unstable 🙅‍♂️ . 
2. We cannot use `?` in `map`.

Map creates a new type anyway so the effect is the same (albeit more verbose).